Week 2

k-ye0415 opened the pull request:
2주차: "My garden"과 "Plant list"의 리스트를 구현해봅시다.

- view model에서 서버에서 받아온 데이터가 아닌 그냥 hard-coding된 데이터(mock이라고 합니다)를 view로 전달하도록 설계합니다.
- view에서는 view model의 상태를 확인해서 리스트 레이아웃을 구성합니다.

## 📝 숙제
Plant 클래스 생성.
임시의 Plant 클래스 타입의 Sample List 를 생성하여 PlantList 와 MyGarend View 에 리스트 구성.


## 📚 얼렁뚱땅 보기만 했지 처음 써봅니다
`MutableStateFlow`
변경 가능한(mutable) 상태(state) 흐름(flow) 상태를 업데이트 할 수 있는 흐름이다.
최신 값을 유지하며, 값이 변경될 때 UI가 자동으로 업데이트 된다.
코루틴을 사용하여 비동기적으로 데이터 흐름을 관리할 수 있다.

`StateFlow`
MutableStateFlow 를 통해 초기 값을 가지는 "읽기 전용" 이다.
값이 변경될 때 자동으로 감지하고 UI를 업데이트하는데, 이때 `collectAsState()`를 사용한다.

`collectAsState` 
수집(collect) 변환(as) 상태(state) stateFlow의 데이터를 수집하여 상태를 특정 타입으로 변환한다.
해당 메소드가 구현되어있는 곳을 보니 여기서 코루틴을 사용한다. 즉, Compose 에서 자동으로 관리를 한다.
Flow 에서 값을 수집하여 Compose UI에서 State 로 변환하는 역할.

---

20250303
### 상태를 관리할 수 있는 것들을 한 곳에 모아놓고 관리한다.
<table style="width: 100%; border-collapse: collapse;">
    <tr>
        <td style="width: 50%; border: 1px solid black; padding: 10px;">이전</td>
        <td style="width: 50%; border: 1px solid black; padding: 10px;">이후</td>
    </tr>
    <tr>
        <td style="width: 50%; border: 1px solid black; padding: 10px;">
            <pre><code>
data class Plant(
    val id: Int,    
    val name: String,    
    val description: String,    
    val growZoneNumber: Int,    
    val wateringInterval: Int,    
    val imageUrl: String = ""
)
            </code></pre>
        </td>
        <td style="width: 50%; border: 1px solid black; padding: 10px;">
            <pre><code>
data class Plant(
    val id: Int,
    val name: String,
    val description: String,
    val growZoneNumber: Int,
    val wateringIntervalInDays: Int, // 단위 명확하게
    val imageUrl: String, // 기본 값 제거
    val addTimestamp: Long, // Long 타입으로 선언
    var lastWateredTimestamp: Long // Long 타입으로 선언
)
            </code></pre>
        </td>
    </tr>
</table>

`wateringIntervalInDays` : 누가 봐도 해당 변수 (또는 함수) 가 무엇을 하는지 파악하기 쉽게 작명할 수 있는 습관을 기르자!  
`imageUrl` : "나중에 추가될꺼니까~" 라는 안일하고 아주 짧은 생각으로 했던 것이 이렇게 후폭풍이 오네요! 신경쓰겠습니다!  
`addTimestamp`, `lastWateredTimestamp` : 상태를 관리할 수 있는 데이터인데 생각을 못했다니, 아직은 한참 멀었네요ㅠ.. 또한, 해당 변수들을 추가하면서 Long 타입으로 선언 된 변수들은 UI에 그려줄 때 확장 함수를 사용할 수 있도록 하고 싶었습니다.  

```kotlin
private val systemZoneId = ZoneId.systemDefault()

fun Long.formatAsDate(format: String = "MMM dd, yyyy"): String {
    val formatter = DateTimeFormatter.ofPattern(format)
    return Instant.ofEpochMilli(this)
        .atZone(systemZoneId)
        .toLocalDate()
        .format(formatter)
}
....
MyGardenScreen.kt
        ...
        // 추가된 날짜
        Text(text = "Planted")
        Text(text = plant.addTimestamp.formatAsDate()) // 사실 확장함수가 무척 매력적으로 보였습니다. 그래서 한번 꼭 써보고 싶었습니다. 적절한 것이 아닐지라도 혼나면서 배우겠습니다.
        ...
```


### Coil 라이브러리
랜덤 이미지를 가져오려고 했을 때 Unsplash 를 사용하려 했으나, 404 에러가 계속 나기에 찾아보니 api 를 추가해서 사용한다고 하기에, 다음 주에 제대로 진행해봐야 할것 같다.
그래서 다른 picsum 이라는 링크에서 랜덤 이미지를 불러와 적용하였다.


### 다중줄 문자열
- trimIndent 가 뭔데 자동으로 만들어주지?
  - 코드에서 가독성을 위해 들여쓰기 한 부분을 실제 사용 부분에서는 제거하는 함수이다.
- 자매품 trimMargin 은 뭘까?
  - 일반 책에서 보면 새로운 단락이 시작 될때 첫 글자 앞에 들여쓰기가 된다. 그런 것을 표현하기 위해 "|" 를 사용하여 적절한 들여쓰기를 표현할 수 있다.

-------
  
20250304

### 시간 관련 변수 Instant 타입으로 변경 및 변수명 수정
멘토링 시작전의 나는 아마 시간 관련 변수 타입을 "String"으로 때려박았을 것이다. 
"어차피~ UI에 그릴 때 String으로 할꺼니까" 라는 생각으로.
하지만 멘토링을 하면서 타입에 대해 알게되었고, 타입의 장단점과 사용성 등을 더 깊게 파보면서 나 나름대로의 생각으로 
"현재로써는 String 은 UI에서만 사용할 것이고, 해당 변수들은 시간 관련된 것이니까 `Long`으로 저장하고 내가 써보고 싶던 확장함수 써봐야지!" 
라는 생각을 가지고 구현했다.
하지만, 멘토님께서 한번 더 깊게 생각하여 줄일 수 있는 부분을 줄이게끔 피드백을 주셨다.
솔직히 놀라웠다.(멘토님께서는 당연하시겠지만..😅)
내가 결정한 것이 최선인지에 대해 계속 고민하는 습관을 들여야겟다.

### Remove codebase-v2.xml from repository
Git 또한 얼렁뚱땅 사용해봤지 잘 몰라서 큰일이다.. 틈틈히 공부해야겠다.

Commit:
PlantList mock data 정의 및 view 전달

Commit:
MyGarden mock data 정의 및 view 전달

Commit:
PlantDetail mock data 정의 및 view 전달

jerry-f-lab commented on Sunflower/.idea/easycode/codebase-v2.xml:
```diff
@@ -0,0 +1,6 @@
+  <component name="com.obiscr.chatgpt.settings.EasyCodeState">
+    <option name="projectFiles" value="$PROJECT_DIR$/app/src/androidTest/java/com/jin/sunflower/ExampleInstrumentedTest.kt;/Users/yejin/Documents/study/sunflower/sunflower-jin/Sunflower/app/src/main/java/com/jin/sunflower/core/model/Plant.kt;/Users/yejin/Documents/study/sunflower/sunflower-jin/Sunflower/app/src/main/java/com/jin/sunflower/feature/main/MainScreen.kt;/Users/yejin/Documents/study/su ...
+  </component>
+</project>
```
이 파일은 .gitignore에 넣어서 repository에서는 제외해주세요.

k-ye0415 replied:
넵 알겠습니다.

jerry-f-lab replied:
파일은 git rm으로 지워주세요

k-ye0415 replied:
넵!

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/model/Plant.kt:
```diff
@@ -0,0 +1,10 @@
+    val description: String,
+    val growZoneNumber: Int,
+    val wateringInterval: Int,
+    val imageUrl: String = ""
```
default argument는 거의 모든 경우에 악(evil)입니다. 일단, default argument없이 해보는건 어떠세요?

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/feature/mygarden/MyGardenScreen.kt:
```diff
@@ -1,21 +1,101 @@
+
+        Spacer(Modifier.height(20.dp))
+        // 추가된 날짜
+        Text(text = "Planted")
+        Text(text = "20250301")
+
+        Spacer(Modifier.height(20.dp))
+        // 마지막 물 준날 + 물 주어야 하는 빈도
+        Text(text = "Last Watered")
+        Text(text = "20250301")
```
이 날짜 정보들도 `Plant` class에 넣어볼까요?

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/feature/mygarden/MyGardenScreen.kt:
```diff
@@ -1,21 +1,101 @@
+                .background(color = Color.Gray),
+            contentAlignment = Alignment.Center
+        ) {
+            Text(text = plant.imageUrl)
```
이미지 utility로 coil을 사용하겠다고 하셨던가요? 실제 인터넷에 떠도는 이미지의 URL을 plant mock에 넣고, coil을 이용해서 그려보는 것은 어떠세요?

https://developer.android.com/develop/ui/compose/graphics/images/loading?hl=ko

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/feature/mygarden/MyGardenScreen.kt:
```diff
@@ -1,21 +1,101 @@
+        ) {
+            Text(text = plant.imageUrl)
+        }
+        Text(text = plant.name, fontSize = 18.sp, fontWeight = FontWeight.Bold)
```
[Q] sp와 dp의 차이는 알고 계신가요?

k-ye0415 replied:
sp vs dp
sp 는 Scale-independent Pixels 의 약자로 단말의 사용자 글꼴 크기 접근성을 반영하여 자동 조정된다.
그래서 주로 Text 에 사용하는 단위이다.

dp 는 Density-independent Pixels 의 약자로 단말의 화면(디스플레이)크기에 따라 자동 조정된다.
주로 height, width 등 고정되는 부분에 사용된다.

jerry-f-lab replied:
텍스트에 dp를 사용하는 경우도 있는데 무엇을 위함일까요?

k-ye0415 replied:
dp를 사용하게 된다면 고정된 크기를 가짐으로써 단말의 사용자 글꼴 크기 접근성에 영향을 받지 않습니다.
즉, 글꼴 크기와는 상관없이 고정되어야 하는 텍스트는 dp로 설정하게 됩니다.
![image](https://github.com/user-attachments/assets/17935035-f5c3-407d-b4c7-bc91454cccb0)

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/feature/plantdetail/PlantDetailScreen.kt:
```diff
@@ -1,21 +1,87 @@
+                .background(color = Color.Gray),
+            contentAlignment = Alignment.Center
+        ) {
+            Text(text = plantDetail.imageUrl)
```
여기도 image를 넣어보면 어떨까요?
2주차는 https://ko.react.dev/learn/thinking-in-react 의, step 1, step 2를 중심으로 하는 시간입니다.

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/model/Plant.kt:
```diff
@@ -0,0 +1,10 @@
+    val name: String,
+    val description: String,
+    val growZoneNumber: Int,
+    val wateringInterval: Int,
```
interval은 단위가, 초인지, 분인지, 시간인지, 밀리초인지, 날인지 예측하기가 어렵네요. 이름이 단위를 넣어주는 건 좋은 습관입니다. 예를 들면, `val wateringIntervalInDays`

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/feature/plantdetail/PlantDetailViewModel.kt:
```diff
@@ -1,7 +1,25 @@
+    private val samplePlantDetail = Plant(
+        id = 1,
+        name = "Apple",
+        description = "An apple is a sweet, edible fruit produced by an apple tree (Malus pumila). Apple trees are cultivated worldwide, and are the most widely grown species in the genus Malus. The tree originated in Central Asia, where its wild ancestor, Malus sieversii, is still found today. Apples have been grown for thousands of years in Asia and Europe, and were brought to North America by  ...
+                "Apple trees are large if grown from seed.\n" +
+                "Generally apple cultivars are propagated by grafting onto rootstocks, which control the size of the resulting tree. There are more than 7,500 known cultivars of apples, resulting in a range of desired\n" +
+                "characteristics. Different cultivars are bred for various tastes and uses, including cooking, eating raw and cider production.\n" +
+                "Trees and fruit are prone to a number of fungal, bacterial and pest problems, which can be controlled by a number of organic and non-organic means. In 2010, the fruit's genome was sequenced as part of research on disease control and selective breeding in apple production.\n" +
+                "Worldwide production of apples in 2014 was 84.6 million tonnes, with China accounting for 48% of the total.",
```
https://kotlinlang.org/docs/strings.html#string-literals

이 부분이 참고가 될 수 있을 것 같습니다.

Commit:
gitignore 에 easycode/codebase-v2.xml 추가.

Commit:
Data class 재정의

- default argument 제거
- 추가/마지막 물준 날짜 추가
- 날짜 타입(long) 선언 후 String 으로 변환할 수 있는 확장 함수 정의

Commit:
coil 라이브러리 추가

- coil 라이브러리 추가 및 사용
- manifest 에 internet permission 추가
- unsplash 는 api 추가가 필요함으로 추후 진행.
- 테스트용으로 picsum 랜덤 이미지 사용

Commit:
변수명 재정의

!! 변수 또는 함수만 보고도 무엇을 의미 하는지 파악 할 수 있도록 변수 또는 함수명 정의 하는 습관 가지기 !!

Commit:
다중 줄 문자열

- trimIndent 가 뭔데 자동으로 만들어주지?
  - 코드에서 가독성을 위해 들여쓰기 한 부분을 실제 사용 부분에서는 제거하는 함수이다.
- 자매품 trimMargin 은 뭘까?
  - 일반 책에서 보면 새로운 단락이 시작 될때 첫 글자 앞에 들여쓰기가 된다. 그런 것을 표현하기 위해 "|" 를 사용하여 적절한 들여쓰기를 표현할 수 있다.

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/model/Plant.kt:
```diff
@@ -0,0 +1,12 @@
+    val wateringIntervalInDays: Int,
+    val imageUrl: String,
+    val addTimestamp: Long,
+    var lastWateredTimestamp: Long
```
여기에도 timestamp의 단위가 초인지 밀리초인지 알기 어려운 단점이 있습니다. 어차피 long을 instant로 변환해서 사용할 것이라면 Plant가 Instant type의 addAt, lastWarteredAt을 선언하시는 건 어떨까요?

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/model/Plant.kt:
```diff
@@ -0,0 +1,12 @@
+    val growZoneNumber: Int,
+    val wateringIntervalInDays: Int,
+    val imageUrl: String,
+    val addTimestamp: Long,
```
add라는 일인칭 동사는 명령의 뉘앙스를 가집니다. 그래서 일인칭 동사는 일반적으로 변수의 이름으로는 적절하지 않아요.
특정 시점에 더해졌다고 말하고 싶다면 addedAt, 특정 시점에 더해야 한다면 shouldAddAt은 어떨까요?

Commit:
시간 관련 변수 Instant 타입으로 변경 및 변수명 수정

- 줄일 수 있는 부분이 있는데 왜 못줄였을까?
  - Long -> Instant 로 변경하여 사용하는 부분을 생략할 수 있는 방법이 있었다.
  이게 최선일까에 대해 고민하는 습관을 기르자.
- 변수명 정의
  - 변수가 무엇을 하는 지 파악하기 쉽게 작성하자. (나만 알아볼 수 있는 코드는 나쁜 코드!)

Commit:
Remove codebase-v2.xml from repository

jerry-f-lab reviewed (approved):
LGTM

jerry-f-lab commented:
> 멘토링 시작전의 나는 아마 시간 관련 변수 타입을 "String"으로 때려박았을 것이다.
> "어차피~ UI에 그릴 때 String으로 할꺼니까" 라는 생각으로.

이렇게 생각하고 계셨었다면 이번 기회에 레이어에 대해서 다시 한번 고찰하는 기회가 되었으면 좋겠네요.

시각(time) 정보를 가지고 어떤 포멧팅을 사용해서 어떤 색과 어떤 폰트로 어떤 위치의 그릴 지 결정하는 것은 view의 영역이죠. 즉, view model은 시각 정보를 view에게 전달하면 그만이지 그 시각을 특정 view의 화면 요구 사항에 맞추어 포맷팅까지 해야할 책임은 없다고 봐야할 것 같아요.

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/core/extensions/DateTimeExtensions.kt:
```diff
@@ -0,0 +1,14 @@
+
+private val systemZoneId = ZoneId.systemDefault()
+
+fun Instant.formatAsDate(format: String = "MMM dd, yyyy"): String {
```
`formatAsDate(instant: Instant)`

jerry-f-lab commented on Sunflower/app/src/main/java/com/jin/sunflower/feature/mygarden/MyGardenScreen.kt:
```diff
@@ -1,21 +1,105 @@
+        Spacer(Modifier.height(20.dp))
+        // 마지막 물 준날 + 물 주어야 하는 빈도
+        Text(text = "Last Watered")
+        Text(text = plant.lastWateredAt.formatAsDate())
```
`formatAsDate(plant.lastWateredAt)`
